src/repr/binary_codec.ml

```diff
@@ -0,0 +1,467 @@
+include Binary_codec_intf
+include Binary_codec_intf.Types
+open Type_core
+open Staging
+module Sizer = Size.Sizer
+
+let unsafe_add_bytes b k = k (Bytes.unsafe_to_string b)
+let str = Bytes.unsafe_of_string
+
+let charstring_of_code : int -> string =
+  let tbl =
+    Array.init 256 (fun i -> Bytes.unsafe_to_string (Bytes.make 1 (Char.chr i)))
+  in
+  fun [@inline always] i ->
+    assert (i < 256);
+    Array.unsafe_get tbl i
+
+module Unit = struct
+  let encode () _k = ()
+  let decode _ ofs = (ofs, ()) [@@inline always]
+  let sizer = Sizer.static 0
+end
+
+module Char = struct
+  let encode c k = k (charstring_of_code (Char.code c))
+  let decode buf ofs = (ofs + 1, buf.[ofs]) [@@inline always]
+  let sizer = Sizer.static 1
+end
+
+module Bool = struct
+  let encode b = Char.encode (if b then '\255' else '\000')
+
+  let decode buf ofs =
+    let ofs, c = Char.decode buf ofs in
+    match c with '\000' -> (ofs, false) | _ -> (ofs, true)
```

Comment:
I'm not 100% sure we want this, but it's currently just copied over from the existing implementation. We don't have failures in the other cases, so maybe @samoht did this for consistency / speed.

Reply:
We shouldn't have failures since we're testing decodings from our own encodings but if someone makes a buggy encoding they won't be able to catch it with our decoding since it accepts all the values

Reply:
But I admit that it must be an extremely rare case and I don't see a case where it would be problematic

Reply:
I mean that we don't fail on invalid inputs in the other cases: e.g. the tags for options / variants, invalid `int` RLEs, [impractically large length headers](https://github.com/mirage/repr/issues/7). I think there's a case for being more defensive systematically, but we should try to be consistent in what guarantees we provide (and optimise for performance otherwise).

Reply:
Since we're the ones using it in the first place, I agree with you ;-)